Enable singing of single-file bundles

swaroop-sridhar opened the pull request:
This change removes the requirement that the bundle-signature
be found at the end of the file, so that tools like signtool
can append their custom information at the end of the file.

The layout of .net core bundles was

AppHost Code
Bundled Files
Manifest header
Manifest entries
Bundle footer
<EOF>

The bundle-footer is now removed, its functionality is served by a
special global in apphost (Bundle-Placeholder) similar to the way
APP-DLL name is encoded in the AppHost.

The normal unbundled apphost has a default value for the bundle-placeholder.
If the apphost is a single-file bundle, the bundler tool (in HostModel) updates
the value of the variable on disk to a different value, and encodes the location
of the manifest header in it.

This way, there is no requirement that the bundle meta-data must remain at
the end of the file.

This change also simplifies the test for whether an apphost is a bundle:
we only check a value in the loaded binary, rather than performing file reads.

swaroop-sridhar committed:
Enable singing of bundles

This change removes the requirement that the bundle-signature
be found at the end of the file, so that tools like signtool
can append their custom information at the end of the file.

The layout of .net core bundles was

AppHost Code
Bundled Files
Manifest header
Manifest entries
Bundle footer
<EOF>

The bundle-footer is now removed, its functionality is served by a
special global in apphost (Bundle-Placeholder) similar to the way
APP-DLL name is encoded in the AppHost.

The normal unbundled apphost has a default value for the bundle-placeholder.
If the apphost is a single-file bundle, the bundler tool (in HostModel) updates
the value of the variable on disk to a different value, and encodes the location
of the manifest header in it.

This way, there is no requirement that the bundle meta-data must remain at
the end of the file.

This change also simplifies the test for whether an apphost is a bundle:
we only check a value in the loaded binary, rather than performing file reads.

AustinWise commented on src/managed/Microsoft.NET.HostModel/AppHost/HostWriter.cs:
```diff
@@ -88,37 +85,66 @@ public static class HostWriter
 
-            // Copy apphost to destination path so it inherits the same attributes/permissions.
-            File.Copy(appHostSourceFilePath, appHostDestinationFilePath, overwrite: true);
+                    bundleHeaderOffset = accessor.ReadInt64(position - sizeof(Int64));
```
Perhaps this method should if `bundleHeaderOffset` is zero, like `marker_t::is_bundle()` does.

swaroop-sridhar replied:
@AustinWise, change is in this commit: https://github.com/dotnet/core-setup/pull/6885/commits/75f19262c4d87f3ed75e1b08e2e7c26fd304cd23

AustinWise replied:
Thanks, that should make for a more clear error message.

lpereira commented on src/corehost/cli/apphost/bundle/bundle_runner.cpp:
```diff
@@ -166,28 +167,9 @@ void bundle_runner_t::reopen_host_for_reading()
-    seek(m_bundle_stream, header_offset, SEEK_SET);
+    seek(m_bundle_stream, marker_t::header_offset(), SEEK_SET);
 
     manifest_header_t* header = manifest_header_t::read(m_bundle_stream);
```
Since `header` is not stored anywhere, and it's allocated in `manifest_header_t`, the memory leaks.  It might be a better idea to allocate on the stack here and have `manifest_header_t::read()` take a point to it as a parameter.

swaroop-sridhar replied:
@lpereira  please see this commit
https://github.com/dotnet/core-setup/pull/6885/commits/bd597daecedfba1995832d66d720b37d35b08c39

swaroop-sridhar committed:
Address feedback by AustinWise

swaroop-sridhar committed:
Fix a build error in test runs

Changes a storage class so that constants are held without warning.

swaroop-sridhar committed:
Address comments by @lpereira

Store bundle_header* in bundle_runner so that it is destructed
after bundle processing (and eliminates redundant storage of
bundle_header contents in bundle_runner).

swaroop-sridhar committed:
Add a test case

swaroop-sridhar committed:
Some additional cleanup

vitek-karas reviewed (approved):
LGTM

lpereira commented on src/corehost/cli/apphost/bundle/bundle_runner.cpp:
```diff
@@ -302,22 +273,11 @@ StatusCode bundle_runner_t::extract()
-        }
-        process_manifest_header(manifest_header_offset);
+        // Read the bundle header
+        seek(m_bundle_stream, marker_t::header_offset(), SEEK_SET);
```
It's often a good idea to replace `seek()`+`read()` with `pread()` if supported by the system.  This way, if reading with an invalid offset, the `pread()` call will fail.

swaroop-sridhar replied:
@vitek-karas  the `seek()` function above already checks for failure, and throws the bundleIOerror exception.

swaroop-sridhar replied:
@lpereira  I agree that the PAL can be improved by adding `pread()`. 
But I think this code will soon change to use memory-mapping of the whole bundle file instead of seek-read-write, etc. So, I left it as-is for now.

lpereira commented on src/corehost/cli/apphost/bundle/header.h:
```diff
@@ -0,0 +1,51 @@
+#pragma pack(pop)
+        pal::string_t m_bundle_id;
+
+        static const uint32_t m_current_major_version = 0;
```
If it's static, should it be prefixed with `m_`?

swaroop-sridhar replied:
No private statics should be `s_`, but this field can just be a const; so I'll remove the annotations.

lpereira commented on src/corehost/cli/apphost/bundle/bundle_runner.h:
```diff
@@ -50,10 +52,9 @@ namespace bundle
         void extract_file(file_entry_t* entry);
 
         FILE* m_bundle_stream;
+        header_t* m_header;
```
What is destroying `m_header`?  The memory leak is still here.  A few options here, in order from the most preferred to the least preferred:

- Declare this inline as `header_t m_header`. No need to allocate a new struct from the heap. Goes away as soon as `bundle_runner_t` goes away.
- Make this a `std::unique_ptr<header_t>` instead.
- Have a `delete m_header` in `bundle_runner_t::~bundle_runner_t()`.

swaroop-sridhar replied:
I added a destructor to the `bundle_runner_t`.
I did not inline the manifest, because when files are directly processed from the bundle, the manifest will survive beyond the `bundle_runner`  object.
I also wanted to keep the interface of `header_t` similar to `manifest_t` and the managed code in `HostModel`. Thanks.

swaroop-sridhar commented on src/corehost/cli/apphost/bundle/header.cpp:
```diff
@@ -0,0 +1,43 @@
+
+    // bundle_id is a component of the extraction path
+    size_t bundle_id_length = 
+        bundle_runner_t::get_path_length(header->m_data.bundle_id_length_byte_1, stream);
```
`bundle_id` is actually a part of a path (that is, it should be path-compatible, and conform to length limitation). The `get_path_length` method only reads two bytes in a length encoding (because of `MAX_PATH`) restriction. So, that's why I had called it `get_path_length` instead of `get_string_length` 

I can rewrite the function more generically, call it `read_string_length` and do the `MAX_PATH` check elsewhere -- but it doesn't serve much purpose other than the symmetry with `read_string()` function.

vitek-karas replied:
OK - it was just a nit anyway... :-)

swaroop-sridhar committed:
Fix a constant naming

swaroop-sridhar committed:
Rewrite initialization of placeholder

To fix a build error on linux.

swaroop-sridhar committed:
Change version to 1.0

swaroop-sridhar committed:
Add destructor for bundle_runner

swaroop-sridhar committed:
Update Error codes

Leave an error code unused (rather than rewriting) to fix a failure
in the lab jobs

Also fix a spacing

swaroop-sridhar committed:
Make the bundle placeholder volatile

The static placeholder array was optimized out in OSX release builds.
This caused tests in release builds to fail with PlaceholderNotFound exception,
whereas the debug builds passed.

swaroop-sridhar commented:
The win-x64 debug failure says `System.IO.IOException : There is not enough space on the disk : 'D:\a\1\s\bin\tests\win-x64.Debug\8\StandaloneApp\publish\StandaloneApp.exe'` 
The error happens when the fixture is cloned for use, likely to be a testing issue.

swaroop-sridhar committed:
Temporarily disable TestWithAdditionalContentAfterBundleMetadata

Disable this test since it is causing other tests to fail because of
 System.IO.IOException : There is not enough space on the disk

This looks like an infrastructure issue, which I will follow up.

dagood commented:
I hit the `There is not enough space on the disk` errors in my work to migrate to Arcade, too, and moving off the tiny hosted agents onto the Helix queues worked for me: https://github.com/dotnet/core-setup/commit/cbe76ff2c47e20accd927efa65abcf3cdf514cf3. IMO feel free to do so here.

swaroop-sridhar commented:
Thanks @dagood ; I'll try that (in a separate change).

swaroop-sridhar commented:
The Build_Windows_x64 debug leg passed with the test disabled.
A few more legs are waiting for machines to run on, but have previously passed. 
https://dev.azure.com/dnceng/public/_build/results?buildId=239197 

So, I'll merge the change, so that the fix is available in the upcoming preview.

nxtn commented on src/corehost/cli/apphost/bundle/marker.cpp:
```diff
@@ -0,0 +1,33 @@
+int64_t marker_t::header_offset()
+{
+    // Contains the bundle_placeholder default value at compile time.
+    // If this is a single-file bundle, the last 8 bytes are replaced 
```
I believe it is the **first** 8 bytes that are replaced?

vitek-karas replied:
Good catch - thanks a lot!
https://github.com/dotnet/runtime/pull/44874
